source/common/local_reply/local_reply.cc

```diff
@@ -0,0 +1,161 @@
+public:
+  ResponseMapper(
+      const envoy::extensions::filters::network::http_connection_manager::v3::ResponseMapper&
+          config,
+      Server::Configuration::FactoryContext& context)
+      : filter_(AccessLog::FilterFactory::fromProto(config.filter(), context.runtime(),
+                                                    context.random(),
+                                                    context.messageValidationVisitor())) {
+    if (config.has_status_code()) {
+      status_code_ = static_cast<Http::Code>(config.status_code().value());
+    }
+    if (config.has_body()) {
+      body_ = Config::DataSource::read(config.body(), true, context.api());
+    }
+
+    if (config.has_body_format()) {
+      body_formatter_ = std::make_unique<BodyFormatter>(config.body_format());
+    }
+  }
+
+  bool matchAndRewrite(const Http::RequestHeaderMap& request_headers,
+                       Http::ResponseHeaderMap& response_headers,
+                       const Http::ResponseTrailerMap& response_trailers,
+                       StreamInfo::StreamInfoImpl& stream_info, Http::Code& code, std::string& body,
+                       BodyFormatter*& final_formatter) const {
+    // If not matched, just bail out.
+    if (!filter_->evaluate(stream_info, request_headers, response_headers, response_trailers)) {
+      return false;
+    }
+
+    if (body_.has_value()) {
+      body = body_.value();
+    }
+
+    if (status_code_.has_value() && code != status_code_.value()) {
+      code = status_code_.value();
+      response_headers.setReferenceStatus(std::to_string(enumToInt(code)));
+      stream_info.response_code_ = static_cast<uint32_t>(code);
+    }
+
+    if (body_formatter_) {
+      final_formatter = body_formatter_.get();
+    }
+    return true;
+  }
+
+private:
+  const AccessLog::FilterPtr filter_;
+  absl::optional<Http::Code> status_code_;
+  absl::optional<std::string> body_;
+  BodyFormatterPtr body_formatter_;
+};
+
+using ResponseMapperPtr = std::unique_ptr<ResponseMapper>;
+
+class LocalReplyImpl : public LocalReply {
+public:
+  LocalReplyImpl() : body_formatter_(std::make_unique<BodyFormatter>()) {}
+
+  LocalReplyImpl(
+      const envoy::extensions::filters::network::http_connection_manager::v3::LocalReplyConfig&
+          config,
+      Server::Configuration::FactoryContext& context)
+      : body_formatter_(config.has_body_format()
+                            ? std::make_unique<BodyFormatter>(config.body_format())
+                            : std::make_unique<BodyFormatter>()) {
+    for (const auto& mapper : config.mappers()) {
+      mappers_.emplace_back(std::make_unique<ResponseMapper>(mapper, context));
+    }
+  }
+
+  void rewrite(const Http::RequestHeaderMap* request_headers,
+               StreamInfo::StreamInfoImpl& stream_info, Http::Code& code, std::string& body,
+               absl::string_view& content_type) const override {
+    Http::RequestHeaderMapImpl empty_request_headers;
```

Comment:
This still requires an allocation on every rewrite call. Can you define some static/constant  empty headers to use for the default case?

Reply:
Done